File: Data Structures & Algorithms/meeting-schedule/submission-7.java
/**
 * Definition of Interval:
 * public class Interval {
 *     public int start, end;
 *     public Interval(int start, int end) {
 *         this.start = start;
 *         this.end = end;
 *     }
 * }
 */

class Solution {
    public boolean canAttendMeetings(List<Interval> intervals) {
        boolean ret = true;
        if(intervals.size() <= 1){
            return true;
        }
        intervals.sort(Comparator.comparingInt(interval -> interval.start));
        Interval first = intervals.get(0);
        for(int i = 1; i < intervals.size(); i++){
            if(intervals.get(i).start < first.end){
                return false;
            }else{
                first.start = intervals.get(i).start;
                first.end = intervals.get(i).end;
            }   
        }

        return ret;
    }
}

File: Data Structures & Algorithms/meeting-schedule-ii/submission-7.js
/**
 * Definition of Interval:
 * class Interval {
 *   constructor(start, end) {
 *     this.start = start;
 *     this.end = end;
 *   }
 * }
 */

class Solution {
    /**
     * @param {Interval[]} intervals
     * @returns {number}
     */
    minMeetingRooms(intervals) {
        let start = [], end = []
        for(let int of intervals){
            start.push(int.start)
            end.push(int.end)
        }
        start.sort((a,b) => a - b)
        end.sort((a,b) => a - b)
        let left = 0, right = 0;
        let maxRooms = 0;
        let curRooms = 0;
        while(left < start.length && right < end.length){
            if(start[left] < end[right]){
                curRooms += 1
                left += 1
            }else{
                curRooms -= 1
                right += 1
            }
            maxRooms = Math.max(maxRooms, curRooms)
        }
        return maxRooms
    }
}
